x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items_details.ts

```diff
@@ -0,0 +1,172 @@
+import { AGENT_NAME, AGENT_VERSION, SERVICE_ENVIRONMENT, SERVICE_LANGUAGE_NAME, SERVICE_NAME, SERVICE_NODE_NAME } from "@kbn/apm-plugin/common/elasticsearch_fieldnames";
+import { environmentQuery } from "@kbn/apm-plugin/common/utils/environment_query";
+import { AgentName } from "@kbn/apm-plugin/typings/es_schemas/ui/fields/agent";
+import { ProcessorEvent } from "@kbn/observability-plugin/common/processor_event";
+import { kqlQuery, rangeQuery } from "@kbn/observability-plugin/server/utils/queries";
+import { RandomSampler } from "../../lib/helpers/get_random_sampler";
+import { AgenItemsSetup } from "./get_agents_items";
+
+export function serviceNameQuery(
+  serviceName?: string
+): QueryDslQueryContainer[] {
+  if (!serviceName) {
+    return [];
+  }
+
+  return [{ term: { [SERVICE_NAME]: serviceName } }];
+}
+
+export function agentLanguageQuery(
+  agentName?: string
+): QueryDslQueryContainer[] {
+  if (!agentName) {
+    return [];
+  }
+
+  return [{ term: { [SERVICE_LANGUAGE_NAME]: agentName } }];
+}
+
+interface AggregationParams {
+  environment: string;
+  serviceName?: string;
+  agentLanguage?: string;
+  kuery: string;
+  setup: AgenItemsSetup;
+  maxNumServices: number;
+  start: number;
+  end: number;
+  randomSampler: RandomSampler;
+}
+
+export async function getAgentsDetails({
+  environment,
+  agentLanguage,
+  serviceName,
+  kuery,
+  setup,
+  maxNumServices,
+  start,
+  end,
+  randomSampler,
+}: AggregationParams) {
+  const { apmEventClient } = setup;
+
+  const response = await apmEventClient.search(
+    'get_agent_details',
+    {
+      apm: {
+        events: [ProcessorEvent.metric],
+      },
+      body: {
+        track_total_hits: false,
+        size: 0,
+        query: {
+          bool: {
+            filter: [
+              {
+                exists: {
+                  field: AGENT_NAME,
+                }
+              },
+              {
+                exists: {
+                  field: AGENT_VERSION,
+                }
+              },
+              {
+                exists: {
+                  field: SERVICE_NODE_NAME,
+                }
+              },
+              ...rangeQuery(start, end),
+              ...environmentQuery(environment),
+              ...kqlQuery(kuery),
+              ...serviceNameQuery(serviceName),
+              ...agentLanguageQuery(agentLanguage),
+            ],
+          },
+        },
+        aggs: {
+          sample: {
+            random_sampler: randomSampler,
+            aggs: {
+              services: {
+                terms: {
+                  field: SERVICE_NAME,
+                  size: maxNumServices,
+                },
+                aggs: {
+                  instances: {
+                    cardinality: {
+                      field: SERVICE_NODE_NAME,
+                    }
+                  },
+                  serviceNodes: {
+                    terms: {
+                      field: SERVICE_NODE_NAME,
```

Comment:
Is the terms agg on `SERVICE_NODE_NAME` needed in this case? Do environments have to be a sub agg of node name?

Reply:
You are right, no need to aggr by environment inside serviceNodes at this stage, this needs to be done whenever we are getting individual instances